What is a Module in Python?
A module in Python is a file that contains Python code (functions, classes, variables, or even runnable code) and is used to organize and reuse code efficiently.

A module is simply a .py file that can be imported and used in other Python programs.

Modules help keep the code modular, readable, and maintainable.

Python has built-in modules (like math, random, os) and also allows users to create custom modules.

Types of Modules in Python
1. Built-in Modules (Standard Library)
Pre-installed modules in Python.
Example: math, random, os, sys
Example usage:

In [ ]:
import math
print(math.sqrt(25))  # Output: 5.0

2. User-Defined Modules (Custom Modules)
Any Python file (.py) you create can be used as a module.
Example:
Create a file called mymodule.py: (use VSCode/Cursor on local computer)

In [ ]:
def add(a, b):
    return a + b

# import and use it another script

In [ ]:
import mymodule
print(mymodule.add(5, 3))  # Output: 8

3. External Modules (Third-party Libraries)
Installed via pip (pip install module_name).
Example: numpy, pandas, requests
Example usage:

In [ ]:
pip install requests


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [4]:
import requests
response = requests.get("https://www.example.com")
print(response.status_code)

200


How to Import a Module in Python?
Python provides several ways to import modules:

1. Basic Import

In [6]:
import math 
print(math.pi)

3.141592653589793


2. Import with Alias (as)

In [ ]:
import numpy as np
print(np.array([1, 2, 3]))

3. Import Specific Functions or Variables (from ... import ...)

In [7]:
from math import sqrt, pi
print(sqrt(16))
print(pi)

4.0
3.141592653589793


In [8]:
from math import sqrt as s, pi as p

print(s(16))
print(p )

4.0
3.141592653589793


 4. Import Everything (from module import *) (Not recommended for large modules)**  

In [9]:
from math import * # wild card
print(sin(0))  # Output: 0.0

0.0


In [10]:
# Case 1: 'import math' (Lazy-loading)
import math  # Only loads the module object
# No extra memory used until `math.sqrt()` is called.

# Case 2: 'from math import *' (Eager-loading)
from math import *  # Loads ALL names (pi, sin, cos, sqrt, ...)
# Memory usage increases even if you never use `pi` or `sin`.

What’s Happening in This Namespace Overlap?
You’re executing:

In [11]:
from math import *
from numpy import *
print(pi)  # Which `pi` is being printed? math.pi or numpy.pi?

3.141592653589793


This is a classic namespace collision caused by wildcard imports (from ... import *). Here’s what Python does:

How Python Resolves the Conflict
from math import * → Dumps all of math’s names (e.g., pi, sin, sqrt) into the global namespace.
from numpy import * → Dumps all of numpy’s names into the same namespace, overwriting any duplicates.
Which pi Wins?
The last imported pi takes precedence (in this case, numpy.pi).
So, print(pi) will output numpy.pi (≈ 3.141592653589793), not math.pi.

Advantages of Using Modules
✔ Code Reusability – Write once, use anywhere.

✔ Organization – Keep related functions together.

✔ Namespace Management – Prevents variable conflicts.

✔ Faster Development – Use existing libraries instead of writing everything from scratch.




Understanding Functions in Python
A Python function is a block of organized, reusable code that is used to perform a single, related action. Functions provide better modularity for your application and a high degree of code reusing.

In [12]:
# This is a global function because it's defined at the top level of the module.
def my_function():
  print("Hello! World")

# The function can be called from anywhere in the module.
my_function()

Hello! World


A global function in Python is a function that’s defined in the main body of a module, rather than inside another function or class. This means that the function is available throughout the module, and if the module is imported into another file, the function can be accessed from there as well.

Key Points:
Scope:
Global functions have a module-level scope. They can be called by any code within that module, and if imported, they can be used elsewhere too.
Usage:
Global functions are typically used to perform tasks that don't depend on a specific object's state. They’re ideal for utility functions, helper functions, or any code that can be reused in various parts of your program.

Types of Python Functions
Python provides the following types of functions −

1) Built-in functions

Python's standard library includes number of built-in functions. Some of Python's built-in functions are print(), int(), len(), sum(), etc. These functions are always available, as they are loaded into computer's memory as soon as you start Python interpreter.
2) Functions defined in built-in modules

The standard library also bundles a number of modules. Each module defines a group of functions. These functions are not readily available. You need to import them into the memory from their respective modules.
3) User-defined functions

In addition to the built-in functions and functions in the built-in modules, you can also create your own functions. These functions are called user-defined functions.


In [13]:
# Built-in functions
print("Hello! World")

Hello! World


In [14]:
# Functions defined in built-in modules
import random
print(random.random())

0.705776517122172


In [ ]:
# User-defined functions
def my_function():
  print("Hello! Operation Badar")

my_function()

Syntax to Define a Python Function 

In [ ]:
def function_name( parameters ):
   "function_docstring"
   function_suite
   return [expression]

In [ ]:
def greetings():
   "This is docstring of greetings function"
   greet = 'Hello World!'
   return greet

message = greetings()
print(message)

Pass by Reference vs Value
Python uses pass by object reference. Immutable objects (e.g. integers) are unchanged, while mutable objects (e.g. lists) are modified. Examples:

Integers: x = 5 remains 5 after modification.
Lists: x = [1, 2, 3] becomes [1, 2, 3, 4] after appending 4.
In this example, x remains unchanged after the modify_value function, because it's an immutable integer. However, lst is modified after the modify_list function, because it's a mutable list.

In [15]:
def modify_value(x):
    x = 10
    print("Within function:", x)

# Immutable object (integer)
x = 5
print("Original:", x)
modify_value(x)
print("After function:", x)

Original: 5
Within function: 10
After function: 5


In [16]:
def modify_list(lst):
    lst.append(4)
    print("Within function: ", lst, " - ID:", id(lst))

# Mutable object (list)
lst = [1, 2, 3]
print("Original:", lst, " - ID:", id(lst))
modify_list(lst)
print("After function:", lst, " - ID:", id(lst))

Original: [1, 2, 3]  - ID: 2265158072000
Within function:  [1, 2, 3, 4]  - ID: 2265158072000
After function: [1, 2, 3, 4]  - ID: 2265158072000


Function Arguments
Function arguments are the values or variables passed into a function when it is called.

In [17]:
def greetings(name):
   "This is docstring of greetings function"
   print ("Hello {}".format(name))
   return

greetings("Ali")
greetings("Omar")
greetings("Usman")

Hello Ali
Hello Omar
Hello Usman


## **Keyword Arguments**

Keyword arguments are related to the function calls. When you use keyword arguments in a function call, the caller identifies the arguments by the parameter name. This allows you to skip arguments or place them out of order because the Python interpreter is able to use the keywords provided to match the values with parameters.

In [18]:
def printinfo( name, age ):
   "This prints a passed info into this function"
   print ("Name: ", name)
   print ("Age ", age)
   return;

# Now you can call printinfo function
printinfo( age=50, name="Arif" )
#printinfo(50, "Arif" )

Name:  Arif
Age  50


In [19]:
def add(x: int,y: int=0) -> float:
   return float(x + y)

print(float(add(10,20)))

print(add(y=50.0, x=2.0)) # type hints are not enforced in Python

print(add(x=5))

30.0
52.0
5.0


* unpacking iterables
In Python, the * operator is used for unpacking iterables (like lists, tuples, or sets) into individual elements. When you use * before a list (or any iterable) in a function call, it unpacks the list and passes its elements as separate positional arguments to the function.

Example:

Sure! Here’s the explanation in Roman English:

Python mein, * operator ek powerful tool hai jo iterables jaise lists, tuples, ya sets ke saath kaam karne ke liye istemal hota hai. Jab aap * ko ek iterable se pehle lagate hain, toh yeh us iterable ke items ko unpack karta hai, aur har item ko alag argument ke roop mein function ko bhejta hai.

### Example:

Maan lijiye, aapke paas ek function hai jo kai arguments leta hai:

```python
def greet(name1, name2, name3):
    print(f"Hello, {name1}, {name2}, and {name3}!")
```

Agar aapke paas names ka ek list hai aur aap is function ko use karna chahte hain:

```python
names = ["Alice", "Bob", "Charlie"]
greet(*names)
```

Yahan, `*names` list ko unpack karta hai, toh yeh waise hi hai jaise aap yeh kahte hain: `greet("Alice", "Bob", "Charlie")`. Phir yeh function print karega:  
`Hello, Alice, Bob, and Charlie!`

### Key Points:
- * operator unpacking ke liye istemal hota hai.
- Yeh iterable ke elements ko alag arguments mein spread karta hai.
- Yeh feature function calls ko clean aur concise banane ke liye bahut useful hai.

Agar aapko aur examples ya explanations chahiye, toh bejhijhak pooch sakte hain!

In [ ]:
def my_sum(*nums):
  print(type(nums),", ", nums)

  return sum(nums)

print("Sum     = ", my_sum(1,2,3,4,5,8,5),"\n")
print("Sum *[] = ", my_sum(*[1,2,3,4,5,8,5]), "\n") # *  unpacking list
print("Sum *() = ", my_sum(*(1,2,3,4,5,8,5))) # *  unpacking tuple


## **Default Arguments**
A default argument is an argument that assumes a default value if a value is not provided in the function call for that argument.


In [20]:
def printinfo( name, age = 35 ):
   "This prints a passed info into this function"
   print ("Name: ", name)
   print ("Age ", age)
   return;

# Now you can call printinfo function
printinfo( age=50, name="Arif" )
printinfo( name="Arif" )

Name:  Arif
Age  50
Name:  Arif
Age  35


Positional-only arguments
Those arguments that can only be specified by their position in the function call is called as Positional-only arguments. They are defined by placing a "/" in the function's parameter list after all positional-only parameters.

Example

In the following example, we have defined two positional-only arguments namely "x" and "y". This method should be called with positional arguments in the order in which the arguments are declared, otherwise, we will get an error.

Bilkul! Yahaan positional-only arguments ko Roman English mein vyakhya di gayi hai:

### Positional-only Arguments

Positional-only arguments wo arguments hote hain jo sirf unki position ke zariye function call mein specify kiye ja sakte hain. Inhe function ki parameter list mein saaf karne ke liye, aap ek "/" ka istemal karte hain, jo yeh indicate karta hai ki uske baad wale arguments sirf positional arguments hain.

### Example:

Maan lijiye, aapne ek function define kiya hai jisme do positional-only arguments hain: `x` aur `y`.

```python
def add(x, y, /):
    return x + y
```

Is function ko aap sirf positional arguments ke saath bulane ki koshish karenge, jaise:

```python
result = add(3, 5)
print(result)  # Output: 8
```

Yahan, `add(3, 5)` sahi hai, kyunki aapne arguments ko unki position ke hisaab se diya hai.

Lekin agar aap try karte hain inhe keyword arguments ke saath specify karne ki, jaise:

```python
result = add(x=3, y=5)
```

Toh yeh error dega, kyunki `x` aur `y` ko sirf positional arguments ke roop mein pass karna zaroori hai.

### Key Points:
- Positional-only arguments ko sirf unki position se hi specify kiya ja sakta hai.
- Yeh arguments "/" ke baad define kiye jaate hain.
- Agar aap inhe keyword arguments se bulane ki koshish karte hain, toh error aayega.

Agar aapko aur examples ya details chahiye, toh poochne mein hichkichaiye mat!

In [21]:
def posFun(x, y, /, z):
    print(x + y + z)

print("Evaluating positional-only arguments: ")
posFun(1, 2, z=3)

# uncomment to see error
#posFun(x=1, y=2, z=3)

Evaluating positional-only arguments: 
6


In [22]:
# Run to see error
posFun(x=1, y=2, z=3)

TypeError: posFun() got some positional-only arguments passed as keyword arguments: 'x, y'

Error
posFun(x=1, y=2, z=3)
This means that arguments before the '/' must be specified by their position in the function call and cannot be passed using keyword arguments.

x and y are declared before the '/', making them positional-only. When you call posFun(x=1, y=2, z=3), you're attempting to pass x and y as keyword arguments, violating this rule and hence the TypeError is raised.

Keyword-only arguments
Those arguments that must be specified by their name while calling the function is known as Keyword-only arguments. They are defined by placing an asterisk ("*") in the function's parameter list before any keyword-only parameters. This type of argument can only be passed to a function as a keyword argument, not a positional argument.

In [23]:
def posFun(*, num1, num2, num3):
    print(num1 * num2 * num3)

print("Evaluating keyword-only arguments: ")
posFun(num1=6, num2=8, num3=5)

posFun(num3=6, num1=8, num2=5)


# TypeError: posFun() takes 0 positional arguments but 3 were given
#posFun(6, 8, 5)

Evaluating keyword-only arguments: 
240
240


Arbitrary or Variable-length Arguments:
YOU CAN AKE MULTIPLE ARGUMENTS
You may need to process a function for more arguments than you specified while defining the function. These arguments are called variable-length arguments and are not named in the function definition, unlike required and default arguments.

Syntax for a function with non-keyword variable arguments is this −


  def functionname([formal_args,] *var_args_tuple ):
    "function_docstring"
    function_suite
    return [expression]


In [25]:
def printinfo( arg1, *vartuple ):
   "This prints a variable passed arguments"
   print ("Output is: ")
   print (arg1)
   for var in vartuple:
      print ("*",var)
   return;

# Now you can call printinfo function
printinfo( 10 )
printinfo( 70, 60, 50, 70, 90 )

Output is: 
10
Output is: 
70
* 60
* 50
* 70
* 90


Python Function with Return Value
The return keyword as the last statement in function definition indicates end of function block, and the program flow goes back to the calling function. Although reduced indent after the last statement in the block also implies return but using explicit return is a good practice.

Along with the flow control, the function can also return value of an expression to the calling function. The value of returned expression can be stored in a variable for further processing.


In [26]:
def add(x,y):
   z=x+y
   return z

a=10
b=20
result = add(a,b)

print ("a = {} b = {} a+b = {}".format(a, b, result))

a = 10 b = 20 a+b = 30


The Anonymous Functions
The functions are called anonymous when they are not declared in the standard manner by using the def keyword. Instead, they are defined using the lambda keyword.

Syntax
The syntax of lambda functions contains only a single statement, which is as follows −

lambda [arg1 [,arg2,.....argn]]:expression

In [ ]:
def add_two(x, y):
  return x + y

my_lambda = lambda x, y:  x + y;

print(my_lambda(1,2))

In [27]:
# prompt: sort by value dictionary using lambda function

my_dict = {"apple": 5, "banana": 2, "cherry": 8, "date": 1}

sorted_dict = dict(sorted(my_dict.items(), key=lambda item: item[1]))

sorted_dict

{'date': 1, 'banana': 2, 'apple': 5, 'cherry': 8}

EXPLANATION OF ABOVE CODE

Bilkul! Yahaan hum dictionary ko value ke hisaab se sort karne ka process Roman English mein samjhenge:

### Dictionary Sorting by Value

Aapke paas ek dictionary hai jiska naam `my_dict` hai:

```python
my_dict = {"apple": 5, "banana": 2, "cherry": 8, "date": 1}
```

Is dictionary mein fruits ke naam keys hain aur unke corresponding numbers unki values hain.

### Sorting Process

1. **`sorted()` Function**: Pehle, hum `sorted()` function ka istemal karte hain, jo kisi bhi iterable (is case mein dictionary) ko sort karta hai. `sorted()` function by default keys ko sort karta hai, lekin hum yahan values ke hisaab se sort karna chahte hain.

2. **`items()` Method**: Jab aap `my_dict.items()` use karte hain, toh yeh dictionary ki saari key-value pairs ko ek list of tuples mein bana deta hai, jaisa:
   ```python
   [('apple', 5), ('banana', 2), ('cherry', 8), ('date', 1)]
   ```

3. **Lambda Function**: Ab, hum ek `key` parameter dete hain `sorted()` function ko, jismein hum ek lambda function istemal karte hain:
   ```python
   key=lambda item: item[1]
   ```

   Yahaan `item` ek tuple hota hai, aur `item[1]` us tuple ki value ko represent karta hai. Iska matlab hai ki hum values ke hisaab se sort karna chahte hain.

4. **Creating Sorted Dictionary**: Sorted result ko `dict()` function ke saath wrapped karte hain taaki hum ek nayi sorted dictionary bana saken:
   ```python
   sorted_dict = dict(sorted(my_dict.items(), key=lambda item: item[1]))
   ```

### Example Output

Agar aap yeh code chalayenge, toh `sorted_dict` kuch is tarah dikhega:

```python
{'date': 1, 'banana': 2, 'apple': 5, 'cherry': 8}
```

Yaani, values ascending order mein sort ki gayi hain.

### Summary

- **`my_dict.items()`**: Dictionary ki key-value pairs ko list of tuples mein convert karta hai.
- **`sorted()`**: Is list ko sort karta hai, jismein lambda function value ke according sorting ko specify karta hai.
- **`dict()`**: Sorted pairs ko wapas ek dictionary mein convert karta hai.

Agar aapko aur koi clarification chahiye ya kuch aur samajhna hai, toh pooch sakte hain!

In [ ]:
# Function definition is here
sum = lambda arg1, arg2: arg1 + arg2;

# Now you can call sum as a function
print ("Value of total : ", sum( 10, 20 ))
print ("Value of total : ", sum( 50, 20 ))

Scope of Variables
All variables in a program may not be accessible at all locations in that program. This depends on where you have declared a variable.

The scope of a variable determines the portion of the program where you can access a particular identifier. There are two basic scopes of variables in Python −

Global variables

Local variables

Global vs. Local variables

Variables that are defined inside a function body have a local scope, and those defined outside have a global scope.

This means that local variables can be accessed only inside the function in which they are declared, whereas global variables can be accessed throughout the program body by all functions. When you call a function, the variables declared inside it are brought into scope.

In [28]:
total = 0; # This is global variable.
# Function definition is here
def sum( arg1, arg2 ):
   # Add both the parameters and return them."
   total = arg1 + arg2; # Here total is local variable.
   print ("Inside the function local total : ", total)
   return total;

# Now you can call sum function
sum( 10, 20 );
print ("Outside the function global total : ", total)

Inside the function local total :  30
Outside the function global total :  0
